src/bjm.c

```diff
@@ -0,0 +1,277 @@
+    if (functionsCount == functionsCapacity) increaseFunctionCapacity();
+    int idx = functionsCount++;
+    functions[idx] = func;
+    jobsByFunc[idx] = zmalloc(sizeof(Joblist));
+    mutexFifoInit(&jobsByFunc[idx]->jobs);
+    jobsByFunc[idx]->func = func;
+    atomicSet(jobsByFunc[idx]->job_count, 0);
+    return idx;
+}
+
+
+/* Pull one job from the active joblists.  Synchronously waits for a job if none available.
+ *  privdata_ptr - returns the caller supplied privdata.
+ *  joblist_ptr  - returns the joblist that the job was taken from.  This is needed by the caller
+ *                 in order to (later) decrement the job_count.
+ * Returns:
+ *  Returns the bjmJobFunc to be called.
+ */
+static bjmJobFunc waitForJob(void **privdata_ptr, Joblist **joblist_ptr) {
+    bjmJobFunc func = NULL;
+
+    mutexFifoLock(&activeJoblists);
+    while (fifoLength(activeJoblists.fifo) == 0) {
+        pthread_cond_wait(&wakeup_cond, &activeJoblists.mutex);
+    }
+
+    Joblist *joblist = fifoPeek(activeJoblists.fifo);
+    func = joblist->func;
+    *joblist_ptr = joblist;
+
+    mutexFifoLock(&joblist->jobs);
+    *privdata_ptr = fifoPop(joblist->jobs.fifo);
+
+    if (fifoLength(joblist->jobs.fifo) == 0) {
+        // No jobs left for this function
+        fifoPop(activeJoblists.fifo);
+    } else if (fifoLength(activeJoblists.fifo) > 1) {
+        // Rotate the joblist for this function to the end
+        fifoPop(activeJoblists.fifo);
+        fifoPush(activeJoblists.fifo, joblist);
+    }
+    // Keep the lock on the individual joblist until it is properly handled in
+    //  the activeJobLists.  Can't have the size changing.
+    mutexFifoUnlock(&joblist->jobs);
+    mutexFifoUnlock(&activeJoblists);
+
+    return func;
+}
+
+
+static void *pthreadFunction(void *arg) {
+    int threadNum = (intptr_t)arg;
+
+    const int MAX_THREAD_NAME = 16;
+    char thread_name[MAX_THREAD_NAME];
+    snprintf(thread_name, MAX_THREAD_NAME, "bjm thread %d", threadNum);
+    redis_set_thread_title(thread_name);
+
+    redisSetCpuAffinity(server.bio_cpulist);
+
+    makeThreadKillable();
+
+    /* Block SIGALRM so only the main thread will receive the watchdog signal. */
+    sigset_t sigset;
+    sigemptyset(&sigset);
+    sigaddset(&sigset, SIGALRM);
+    if (pthread_sigmask(SIG_BLOCK, &sigset, NULL)) {
+        serverLog(LL_WARNING, "Warning: can't mask SIGALRM in BJM thread: %s", strerror(errno));
+    }
+
+    while (1) {
+        void *privdata;
+        Joblist *joblist;
+        bjmJobFunc func = waitForJob(&privdata, &joblist);
+
+        func(privdata);                     // Execute the callback
+        atomicDecr(joblist->job_count, 1);   // Decrement count AFTER callback finishes
+        atomicDecr(queuedJobCount, 1);
+        atomicIncr(processedJobCount, 1);
+    }
+
+    return NULL;
+}
+
+
+// API
+void bjmInit(int numThreads) {
+    if (threadCount == numThreads) return;  // Silently skip to support testing
```

Comment:
Good thing you are doing so much testing :). If we want to better enable testing, we should start allocating the BJM system as its own struct so we can initialize multiple of them (Mentioned this in another comment).